### PR TITLE
Multiple editors for the head and the body

### PR DESCRIPTION
**Pull request by maurizzzio**
This pull request address #73 :

1. Creates new editors controlled by a button group next to the `Rebuild` button, each editor is created the same way `javascript-editor` is built but using another mode, new files: `lib/htmlEditor, ilb/codemirror_modes/`
2. Gists are now created with two additional files: `page-head.html` and `page-body.html`
3. The code for these two new editors is also saved in `localStorage` under `bundleCode`, `headCode` and `bodyCode`

Additional changes that I did which might be excluded from the pull:

1. Replaced `codemirror.css` by the one existing in the `codemirror` module (I think that this file should be left untouched and style overrides should be done on the `theme.css` or `style.css` files)

Update:

Based on previous work now the contents of `package.json` can be modified allowing any version to be bundled

Screenshot:
requirebin

### CODE REVIEW

**Comment by kumavis** [comment]
This is interesting. I've wanted tighter package.json control in the past. On the flipside, right now I can just throw in some requires and it works -- Is the package.json editor here automatically generated or do i need to maintain that as well?

**Comment by kumavis** [comment]
I think a big win of requirebin is being able to open it up, throw in some code, and watch it work. That simplicity and flat learning curve is important to maintain and means we need to fight off feature creep. Requirebin is not meant to be a proper IDE, just a tool for throwing together some simple browserify code together and maybe sharing it.

However i've used it for module usage examples before, and had things break b/c of the lack of package.json control. So you're hitting on something important here.

File: index.js [hunk 1]
```diff
@@ -10,11 +10,17 @@ var detective = require('detective')
 var md5 = require('md5-jkmyers')
 var keydown = require('keydown')
 
+var htmlEditor = require('./lib/htmlEditor')
+var beautify = require('js-beautify').js_beautify;
```
Comment (maxogden) [inline index.js]: is this ever used? I don't see the `beautify` var used anywhere

File: index.js [hunk 2]
```diff
@@ -136,43 +160,74 @@ function initialize() {
...
-    var editor = jsEditor({
+    // javascript editors
+    var bundleEditor = jsEditor({
       container: editorEl,
       lineWrapping: true
     })
+    bundleEditor.name = 'bundle';
+
+    var metaEditor = jsEditor({
+      value: JSON.stringify(packagejson),
```
Comment (maxogden) [inline index.js]: IMO this shold be `JSON.stringify(packagejson, null, '  ')` so that the JSON gets pretty printed

File: index.js [hunk 3]
```diff
@@ -136,43 +160,74 @@ function initialize() {
...
+      container: editorMetaEl,
+      lineWrapping: true
+    })
+    metaEditor.on('afterFocus', function () {
+      metaEditor.setValue(JSON.stringify(packagejson));
```
Comment (maxogden) [inline index.js]: also needs pretty printing here

File: index.js (hunk 1 above)
Comment (maurizzzio) [inline index.js]: sorry about that, it's dead code, thanks for reviewing it :)

**Comment by maxogden** [comment]
+1 to this overall. Under the hood there were already these files already anyway, and this just exposes them. I like that you can still only have 1 JS file.

I have some opinions about the button naming, size and order but I dont wanna bikeshed in here -- I can just change it after merging.

Adding to @kumavis question: I think the behavior of the package.json editor should be:

- When you add a `require('foo')` to the code, it should automatically update the `package.json` (the same way as it does now when you save to gist)
- If you leave the 'package.json' tab and the JSON is invalid, it should revert back to the last working version

**Comment by maurizzzio** [comment]
@maxogden The behavior you mentioned is what I wanted to achieve, whenever the metaEditor is changed the following code is run:

```javascript
var code = metaEditor.editor.getValue();
try {
  window.packagejson = packagejson = JSON.parse(code);
} catch (e) { }
```

This is the only place packagejson is modified aside from place where sandbox emits the `modules` event

**Comment by maxogden** [comment]
@maurizzzio thanks again for the PR! especially the GIF :)

You had broken the Gist publishing (github doesnt allow empty files) but I fixed that, and also changed some other stuff. gonna deploy now

**Comment by maurizzzio** [comment]
@maxogden :smiley: thanks to you for developing this awesome project and sharing it with the world

**Comment by maxogden** [comment]
OK its live! http://requirebin.com/

**Comment by maurizzzio** [comment]
@maxogden I think that there's a problem when trying to save gists, I'm getting `http://localhost:5000/?error=redirect_uri_mismatch&error_description=The+redirect_uri+MUST+match+the+registered+callback+URL+for+this+application.&error_uri=https%3A%2F%2Fdeveloper.github.com%2Fv3%2Foauth%2F%23redirect-uri-mismatch`, are you able to reproduce it?

**Pull request merged**